lib/gollum-lib/git_access.rb

```diff
@@ -156,16 +156,19 @@ def ref_to_sha!(ref)
     # sha - String commit SHA.
     #
     # Returns an Array of BlobEntry instances.
-    def tree!(sha)
-      tree  = @repo.lstree(sha, { :recursive => true })
-      items = []
-      tree.each do |entry|
-        if entry[:type] == 'blob'
-          next if @page_file_dir && !entry[:path].start_with?("#{@page_file_dir}/")
-          items << BlobEntry.new(entry[:sha], entry[:path], entry[:size], entry[:mode])
+    def tree!(sha, path = '/', recursive = true)
+      tree  = @repo.lstree(sha, (Pathname.new(@page_file_dir.to_s) + path).to_s, recursive: recursive )
+      tree.reduce([]) do |accumulator, entry|
+        if !@page_file_dir || entry[:path].start_with?("#{@page_file_dir}/") # guard against directory traversal
+          accumulator << case entry[:type]
+          when 'blob'
+            BlobEntry.new(entry[:sha], entry[:path], entry[:size], entry[:mode])
+          when 'tree'
+            TreeEntry.new(entry[:sha], entry[:path])
+          end
         end
```

Comment:
We could also opt to return blobs and trees in separate Arrays, so callers who need only the blobs can ignore the trees.

Either way, I guess initializing the TreeEntries has minor performance impact when looping into the entire wiki.